### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by zoetrope**
This PR includes implementation and testing.
Building container image and documentation will be a separate PR.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -3,25 +3,25 @@ package cmd
 import (
 	"flag"
 	"fmt"
-	"github.com/cybozu-go/pod-security-admission/hooks"
 	"io/ioutil"
```
Comment (ysksuzuki) [inline cmd/root.go]: `io/ioutil` is deprecated in go 1.16.  Use `os` package instead.

File: hooks/validators/deny_host_namespace.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	p := field.NewPath("spec")
+	var errs field.ErrorList
+	if pod.Spec.HostNetwork {
+		errs = append(errs, field.Forbidden(p.Child("hostNetwork"), "Host network is not allowed to be used"))
+	}
+	if pod.Spec.HostPID {
+		errs = append(errs, field.Forbidden(p.Child("hostPID"), "Host pid is not allowed to be used"))
+	}
+	if pod.Spec.HostIPC {
+		errs = append(errs, field.Forbidden(p.Child("hostIPC"), "Host ips is not allowed to be used"))
```
Comment (tapih) [inline hooks/validators/deny_host_namespace.go]: `ips` should be `ipc`

File: hooks/testdata/privileged/host-ipc.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: host-ipc
+  annotations:
+    test.pod-security.cybozu.com/message: "denied the request: spec.hostIPC: Forbidden: Host ips is not allowed to be used"
```
Comment (tapih) [inline hooks/testdata/privileged/host-ipc.yaml]: `ips` should be `ipc`

File: Makefile [hunk 4]
```diff
@@ -48,45 +49,48 @@ check-generate:
...
-	test -z "$$($(NILERR) $$(go list ./...) 2>&1 | tee /dev/stderr)"
 	go vet ./...
 
 ENVTEST_ASSETS_DIR=$(shell pwd)/testbin
 test: manifests generate ## Run tests.
 	mkdir -p ${ENVTEST_ASSETS_DIR}
-	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.7.0/hack/setup-envtest.sh
-	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test ./... -coverprofile cover.out
+	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v$(CONTROLLER_RUNTIME_VERSION)/hack/setup-envtest.sh
+	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR); go test -v ./... -coverprofile cover.out
```
Comment (tapih) [inline Makefile]: This might get developer's local environments dirty. I prefer to run it inside subshell.

File: hooks/mutate_pod.go [hunk 5]
```diff
@@ -3,23 +3,25 @@ package hooks
...
+	client       client.Client
+	decoder      *admission.Decoder
+	mutatorNames []string
 }
 
 // NewPodMutator creates a webhook handler for Pod.
-func NewPodMutator(c client.Client, dec *admission.Decoder) http.Handler {
-	return &webhook.Admission{Handler: &podMutator{c, dec}}
+func NewPodMutator(c client.Client, dec *admission.Decoder, mutators []string) http.Handler {
+	return &webhook.Admission{Handler: &podMutator{c, dec, mutators}}
```
Comment (tapih) [inline hooks/mutate_pod.go]: Validating the `mutators` arg here might help us notice mistakes in configurations before actually applying pod-related manifests.

File: hooks/mutate_pod.go [hunk 6]
```diff
@@ -30,11 +32,20 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) admissio
 	}
 
 	poPatched := po.DeepCopy()
-
+	for _, name := range m.mutatorNames {
+		mutator, ok := availableMutators[name]
```
Comment (tapih) [inline hooks/mutate_pod.go]: Log for what kind of mutator is applied is helpful to know what happened exactly on Pods.

File: hooks/validate_pod.go [hunk 7]
```diff
@@ -2,25 +2,30 @@ package hooks
...
-	decoder         *admission.Decoder
+	client         client.Client
+	decoder        *admission.Decoder
+	validatorNames []string
 }
 
 // NewPodValidator creates a webhook handler for Pod.
-func NewPodValidator(c client.Client, dec *admission.Decoder) http.Handler {
+func NewPodValidator(c client.Client, dec *admission.Decoder, validators []string) http.Handler {
 	v := &podValidator{
```
Comment (tapih) [inline hooks/validate_pod.go:27]: The same as `mutat_pod.go`

File: hooks/validate_pod.go [hunk 8]
```diff
@@ -32,9 +37,36 @@ func (v *podValidator) Handle(ctx context.Context, req admission.Request) admiss
 		return admission.Errored(http.StatusBadRequest, err)
 	}
 
-	containers := make([]corev1.Container, len(po.Spec.Containers)+len(po.Spec.InitContainers))
-	copy(containers, po.Spec.Containers)
-	copy(containers[len(po.Spec.Containers):], po.Spec.InitContainers)
+	var allErrs field.ErrorList
+	for _, name := range v.validatorNames {
+		validator, ok := availableValidators[name]
```
Comment (tapih) [inline hooks/validate_pod.go]: The same as `mutate_pod.go`.

File: hooks/validators/deny_host_ports.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// DenyHostPorts is a Validator that denies usage of HostPorts
+func DenyHostPorts(ctx context.Context, pod *corev1.Pod) field.ErrorList {
```
Comment (tapih) [inline hooks/validators/deny_host_ports.go]: The host port field in PSP accepts a range of ports to be allowed.
I'm not sure this is needed right now, but please consider it is needed or not.

File: hooks/validators/deny_unsafe_capabilities.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	"NET_RAW",
+	"SETFCAP",
+	"SETGID",
+	"SETPCAP",
+	"SETUID",
+	"SYS_CHROOT",
+}
+
+// DenyUnsafeCapabilities is a Validator that denies adding capabilities beyond the default set
+func DenyUnsafeCapabilities(ctx context.Context, pod *corev1.Pod) field.ErrorList {
```
Comment (tapih) [inline hooks/validators/deny_unsafe_capabilities.go]: The capabilities field in PSP accepts a list of capabilities to be allowed.
I'm not sure this is needed right now, but please consider it is needed or not.

File: hooks/validators/deny_unsafe_capabilities.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	p := field.NewPath("spec")
+	var errs field.ErrorList
+
+	pp := p.Child("containers")
+	for i, co := range pod.Spec.Containers {
+		if co.SecurityContext == nil || co.SecurityContext.Capabilities == nil {
+			continue
+		}
+		for j, add := range co.SecurityContext.Capabilities.Add {
+			if !containsString(defaultCapabilities, string(add)) {
```
Comment (tapih) [inline hooks/validators/deny_unsafe_capabilities.go]: If we use the set pattern with `map[string]struct{}`, we don't need to define `containsString`, but this line of code does not affect the performance of this webhoo badly, so doing it or not i's up to you.

**Review by tapih — CHANGES_REQUESTED** [review]

**Review by tapih — APPROVED** [review]
LGTM

**Pull request merged**